Added implementation of Lagrange Inversion Theorem for analytic functions

avishrivastava11 opened the pull request:
Fixes #14110

added implementation of Lagrange Inversion theorem for analytic functions to calculate expansion of inverse of an analytic function.

Please take a look and suggest improvements.
Thanks.

#### Release Notes

<!-- BEGIN RELEASE NOTES -->
* series
    * Added `sympy.series.series.lagrange_inversion_theorem()`, which calculates the expansion of the inverse of an analytic function.
<!-- END RELEASE NOTES -->

avishrivastava11 committed:
added implementation of Lagrange Inversion Theorem

asmeurer commented:
I think it makes more sense to have this in the series module than the solvers module.

avishrivastava11 committed:
moved code to series from solvers, added tests and initial conditions

asmeurer commented on sympy/series/series.py:
```diff
@@ -10,3 +13,133 @@ def series(expr, x=None, x0=0, n=6, dir="+"):
+
+    try:
+        eq.free_symbols
+    except:
```
Do not use bare except. Check for the exception you are looking for. In this case, every Basic object should have free_symbols, so I don't see a need to do any checks.

avishrivastava11 replied:
Done now, please take a look

asmeurer commented on sympy/series/series.py:
```diff
@@ -10,3 +13,133 @@ def series(expr, x=None, x0=0, n=6, dir="+"):
+    try:
+        eq.free_symbols
+    except:
+        raise AssertionError("The function provided is not analytic i.e. f' = 0. Lagrange's inversion theorem "
```
How does not having the `free_symbols` attribute make it not analytic? Also, `AssertionError` is inappropriate here.

avishrivastava11 replied:
>How does not having the `free_symbols` attribute make it not analytic?

trying `eq.free_symbols` checks for presence of only constants like ` f(x) = 2 `. Since ` f' = 0 ` for all `x` here, this fails the basic condition for lagrange inversion theorem.

As for the `AssertionError`, I'm correcting that now. Thanks for the input.

asmeurer replied:
`eq.free_symbols` would only raise an exception if it doesn't have the attribute. If you want to check if there are no free_symbols use `if not eq.free_symbols`.

asmeurer commented on sympy/series/series.py:
```diff
@@ -10,3 +13,133 @@ def series(expr, x=None, x0=0, n=6, dir="+"):
+    .. [2] http://mathworld.wolfram.com/LagrangeInversionTheorem.html
+    .. [3] http://www.cfm.brown.edu/people/dobrush/am33/Mathematica/lit.html
+    """
+    from sympy.abc import w, x
```
These should be provided as parameters to the function.

asmeurer replied:
If you simply replace this import with making `x` a parameter of the function, the rest of the code will be the same.

asmeurer commented on sympy/series/series.py:
```diff
@@ -10,3 +13,133 @@ def series(expr, x=None, x0=0, n=6, dir="+"):
     return expr.series(x, x0, n, dir)
+
+
+def lagrange_inversion_theorem(eq, no_of_terms = 3):
```
No space around `=` when used for keyword arguments. 

Also, a better name would just be `n`, which is what `series` uses.

avishrivastava11 replied:
Done now, please take a look

asmeurer commented on sympy/series/series.py:
```diff
@@ -10,3 +13,133 @@ def series(expr, x=None, x0=0, n=6, dir="+"):
+    z_diff = diff(eq, w)
+    t = 0
+
+    while (True):
```
No parentheses are needed here.

avishrivastava11 replied:
Done now, please take a look

asmeurer commented on sympy/series/series.py:
```diff
@@ -10,3 +13,133 @@ def series(expr, x=None, x0=0, n=6, dir="+"):
+
+    a = t
+    f_a = z.subs(w, a)
+    sum = 0
```
I would use a different variable name here. `sum` is a builtin name, and overriding it is confusing.

avishrivastava11 replied:
Done now, please take a look

asmeurer commented on sympy/series/series.py:
```diff
@@ -10,3 +13,133 @@ def series(expr, x=None, x0=0, n=6, dir="+"):
+    f_a = z.subs(w, a)
+    sum = 0
+
+    while (n > 0):
```
This could be a for loop.

avishrivastava11 replied:
Done now.

asmeurer commented on sympy/series/series.py:
```diff
@@ -10,3 +13,133 @@ def series(expr, x=None, x0=0, n=6, dir="+"):
+    z = eq
+
+    z_diff = diff(eq, w)
+    t = 0
```
Instead of trying integer values, which might not succeed, this should also be provided as a parameter to the function.

avishrivastava11 replied:
The user might not always be sure of taking the trouble of finding `a` and checking if it's a valid position to be applied for finding the expansion. This is why I took that route. But your suggestion made more sense when I dived into the actual documentation given on Wolfram as the expansion will depend on the point where it is being calculated. So thanks for the input. Have implemented now.

asmeurer commented on sympy/series/series.py:
```diff
@@ -10,3 +13,133 @@ def series(expr, x=None, x0=0, n=6, dir="+"):
+    where f is analytic at a point a and f'(a) is not equal to 0.
+    Then it is possible to invert or solve the equation for w in the form of a series:
+
+    w = a + \sum_{n = 1}^{oo} (g_n*(z - f(a))**n)/n!
```
Use `.. math::` here so that it formats correctly in the docs. I would suggest adding this function to the Sphinx docs so that you can see if it renders correctly.

avishrivastava11 replied:
Yes will do so. That part is yet to be corrected. I wasn't sure about the way the formula is been shown in the documentation (for now). So, this is a point that has to be taken care of surely. I was trying to set all other parts correctly before I do that (It's one of the easier parts, so can be done in the end). Thanks for this. Also, I need to get around with Sphinx docs (I haven't got the good fortune of yet interacting with it). Please do not set this point as ' resolved ' right now.

asmeurer commented on sympy/series/series.py:
```diff
@@ -10,3 +13,133 @@ def series(expr, x=None, x0=0, n=6, dir="+"):
+    sum = 0
+
+    while (n > 0):
+        l = limit(diff(((w - a) / (z - f_a)) ** n, w, n - 1), w, a)
```
Can this derivative be computed in terms of the previous derivative (I'm not sure)?

avishrivastava11 replied:
The previous derivative is calculated to check whether `f'(a)` is zero or not (to check the basic condition where lagrange inversion theorem can be applied). While this derivative is completely different from that derivative. So it can't be computed in terms of the previous derivative.

asmeurer replied:
I mean whether the nth derivative can be computed in terms of the (n-1)th derivative. Since we are increasing n it would make it more efficient. Similar to how in a Taylor series one can use derivative the previous term to compute the next term, instead of differentiating `n` times again.

avishrivastava11 replied:
Oh I see. Yes that's a very good and important point. I should've seen that. I'll try and get back on this, probably tomorrow. Also, I'll check on the Sphinx docs, since that still needs to be done.

avishrivastava11 replied:
@asmeurer I've done this now. Please take a look and recommend if something else can be improved.

maciejskorski replied:
@asmeurer no, I am afraid you cannot simplify this way. You calculate higher and higher derivates of different and different functions each time.

avishrivastava11 commented on sympy/series/series.py:
```diff
@@ -10,3 +13,133 @@ def series(expr, x=None, x0=0, n=6, dir="+"):
+                             "requires that the function be analytic")
+    if len(eq.free_symbols) == 2:
+        if not (x in eq.free_symbols and w in eq.free_symbols):
+            raise AssertionError("The function provided must be in terms of 'x' or 'w'")
```
Done now, please take a look

avishrivastava11 committed:
modified differentiation implementation for calc, tests

avishrivastava11 commented:
@asmeurer  I've modified the implementation so that it doesn't have to calculate the differential every time. Please take a look and if there's anything else that can be improved, let me know.
Also, please help me in "how to add the function to sphinx docs" (from the earlier chat above)

avishrivastava11 committed:
modified examples and tests to fix travis errors

avishrivastava11 commented:
Can some mentor take a look at this now? I think it's good to merge from my side.

avishrivastava11 commented:
@sylee957  Can you take a look at this? (Also, can you help me with why a few of the tests aren't passing? I'm unable to comprehend why, since it just shows a big list of `compiling ' .. '` and `listing ' ... '`)

sylee957 committed:
Docstring fixes

My guesses of the log failing `SyntaxError: invalid escape sequence \s` would be about the formatting of the docstring.
Latex formula `\sum` may interfere with python's escape sequence, so I have changed to the raw docstring.

I have changed the pretty print formula to latex (which may have issues with sphinx as it is sensitive to indentations),
and also fixed the error in one of the formula, from `{\n!}` to `{n!}`

I hope if this works.

sylee957 committed:
Clean unnecessary changes

avishrivastava11 commented:
@sylee957  Thanks for helping me out here :)

@asmeurer  @sylee957  Do you think this can be merged now?

oscarbenjamin commented:
This looks like a good PR that has been sitting for 3 weeks now. Does anyone have any strong opinions on this?

sylee957 commented on sympy/series/series.py:
```diff
@@ -10,3 +13,151 @@ def series(expr, x=None, x0=0, n=6, dir="+"):
+    for i in range(2, n):
+        l = limit(a_2, w, a)
+        result = result + (((z - f_a) ** i) / factorial(i)) * l
+        a_2 = (i+1) * ((w - a / z - f_a) ** i) * a_2
```
Also, the way how `n` binds to the polynomial degree may had changed as in the above example.
Now, it evaluates until it reaches 4 instead of 5.
However, I also think that recent change can be more compatible with the other implementation, `series`.

But `series` also returns a truncated order term appended with the series expansion, like `O((x - 1)**4, (x, 1))`

<img src="https://user-images.githubusercontent.com/34944973/52969175-cadb9080-33f2-11e9-8ef8-b673d7950a1b.png" width="500px">
Can you add the feature to this function?

avishrivastava11 replied:
I'm not sure if I understand what you're trying to say here.
The reason why n=1 always return an integer is because of the formula of Lagrange inversion. It contains the term `a`. This is what is returned as the first term in the expansion and is counted as one of the terms. 
If I've misunderstood what you were trying to say here, please let me know.

oscarbenjamin commented:
@avishrivastava11 this looks like a good contribution to SymPy that I wouldn't want to see wasted. Can you respond to @sylee957's comments?

avishrivastava11 commented:
I had my exams going on for a few days. I'll surely take a look at this in a day or two. Apologies for the delay.

avishrivastava11 commented on sympy/series/series.py:
```diff
@@ -10,3 +13,151 @@ def series(expr, x=None, x0=0, n=6, dir="+"):
+    for i in range(2, n):
+        l = limit(a_2, w, a)
+        result = result + (((z - f_a) ** i) / factorial(i)) * l
+        a_2 = (i+1) * ((w - a / z - f_a) ** i) * a_2
```
@sylee957  First of all, I'm sorry for being so late to reply. I had my exams for the past 2 weeks and was prepping for them. 
Turns out, I didn't implement the derivative correctly (and so the future result which uses the derivative also turned out to be incorrect) as you pointed out. It was correct initially. While trying to make the previously calculated derivative useful, which would have speed ed up the entire process drastically, I miscalculated the way the derivative would be used (my bad here). Turns out that using the previously calculated derivative, is not as easy as it sounds because it is used in products which increase in number drastically as `n` increases. So taking the previously calculated derivative, is not an option.
Fortunately the correct results are displayed now but of course it takes up time.
So far I haven't been able to figure out a way to re-use the derivative. If you can help here it'll be more than great. 
(I've considered using the option of a dictionary of sorts but in that approach, since the derivative is reused a lot of times - in big `n`s, re-iterating the dictionary every single time, will be a costly operation. Also, we'll have to find a way to test at every point of the calculation if it's derivative would be available in the dictionary or not. This again, would be costly.)

avishrivastava11 committed:
corrected derivative implementation

avishrivastava11 committed:
Merge branch 'lagrange_inversion' of https://github.com/avishrivastava11/sympy into lagrange_inversion

avishrivastava11 commented:
@oscarbenjamin  I'd love to receive your views on this as well. Can you help out in speeding the function somehow - maybe re-using the derivatives (see my reply to @sylee957  above)?

oscarbenjamin commented on sympy/series/series.py:
```diff
@@ -10,3 +13,147 @@ def series(expr, x=None, x0=0, n=6, dir="+"):
+
+    w = x
+    eq = eq.subs(x, w)
+    z = eq
```
Do the lines above do anything? It looks to me as if they aren't needed.

avishrivastava11 replied:
@oscarbenjamin  Everywhere I saw, the usage of the formula was in terms of `z` and `w`.  But everyone is comfortable in using `x` as a global variable for inputs. Hence the change as I didn't want anything to go wrong or difficult in terms of understanding the code by someone else or even by me while debugging.

oscarbenjamin replied:
I think saying `w=x` and `z=eq` is fine if you want to use different variables names inside the function. The call to `subs` is just wasting CPU cycles though.

avishrivastava11 replied:
the subs is done to replace `x` with `w` so that the problem can be solved (since everywhere I found the description of the formula in terms of `w` and `z`). If I wouldn't have used `subs` then how would I have used the equation? (Then later, the `w` is replaced back with `x`)

avishrivastava11 replied:
Do you think this can be merged now?

czgdp1807 commented:
@avishrivastava11 It looks like a good work. Would you like to continue this?
@oscarbenjamin @sylee957 @asmeurer What all needs to be done to complete this PR?

maciejskorski commented:
Hi Guys, what's the status of this? I am upvoting too. 

Two cents from me on your workflow: PR contributors get discouraged, when the feedback iterations are too incremental; it would be good to have this tradeoff: less batches (e.g. partial reviews) of more complexity (e.g. concerte TODOs).

maciejskorski commented:
> @avishrivastava11 It looks like a good work. Would you like to continue this?
> @oscarbenjamin @sylee957 @asmeurer What all needs to be done to complete this PR?

We need to correct the implementation and add unit tests.
As of now, the output is not even of the right type (should be polynomial).
But the documentation is quite good.

Continuing in #24243
